### PR TITLE
Multi wildcard support finalization

### PR DESCRIPTION
**Pull request by lrecknagel**
First - thanks for all the efforts already put in this project and also the PR with I`am picking up!

2nd - I hope I have picked up the prior work correctly, please tell me if I take a wrong path here.
(Followed the suggested manual: https://gist.github.com/jsumners/461ef7a64545108635cc437fde112721)

I´ve worked on finishing the multi-wildcard feature #36 with the missing test-coverage.
Also added handling for a `undefined / null` case which I`ve found.

Following checks have been made:
* `npm run cov-ui`
* `npm run test`
* `npm run bench` (see attached report)

[fast-redact-bench-wc_finalization.txt](https://github.com/davidmarkclements/fast-redact/files/7868439/fast-redact-bench-wc_finalization.txt)

Please give me ping if I forgot something - hope we can bring this feature to life!

Best Lucas

### CODE REVIEW

File: lib/modifiers.js [hunk 1]
```diff
@@ -45,7 +45,20 @@ function nestedRestore (arr) {
   const length = arr.length
   for (var i = 0; i < length; i++) {
     const { key, target, value } = arr[i]
-    target[key] = value
+    if (has(target, key)) {
+      target[key] = value
+    }
+    /* istanbul ignore else */
+    if (typeof target === 'object') {
```
Comment (mcollina) [inline lib/modifiers.js:52]: Is it possible that target is not an object? It's not covered by tests anyway

File: lib/modifiers.js [hunk 2]
```diff
@@ -45,7 +45,20 @@ function nestedRestore (arr) {
...
+      const targetKeys = Object.keys(target)
+      for (var j = 0; j < targetKeys.length; j++) {
+        const tKey = targetKeys[j]
+        const subTarget = target[tKey]
```
Comment (mcollina) [inline lib/modifiers.js:56]: what if subtarget is also an object?

File: example/mulit-wildcard-array.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+'use strict'
```
Comment (davidmarkclements) [inline example/mulit-wildcard-array.js:1]: filename typo 👆  (mulit)

File: lib/modifiers.js [hunk 4]
```diff
@@ -67,7 +80,17 @@ function nestedRedact (store, o, path, ns, censor, isCensorFct, censorFctTakesPa
 }
 
 function has (obj, prop) {
-  return Object.prototype.hasOwnProperty.call(obj, prop)
+  return obj !== undefined && obj !== null
+    ? Object.prototype.hasOwnProperty.call(obj, prop)
```
Comment (davidmarkclements) [inline lib/modifiers.js:84]:
```suggestion
    ? 'hasOwn' in Object ? Object.hasOwn(obj, prop) : Object.prototype.hasOwnProperty.call(obj, prop)
```

File: lib/modifiers.js [hunk 5]
```diff
@@ -67,7 +80,17 @@ function nestedRedact (store, o, path, ns, censor, isCensorFct, censorFctTakesPa
...
+    : false
+}
+
+function isIn (obj, prop) {
+  try {
+    return prop in obj
+  } catch (e) {
+    return false
+  }
```
Comment (davidmarkclements) [inline lib/modifiers.js:93]: is this not slower than inlining `typeof b === 'object'  && a in b` wherever it's needed?

**Review by davidmarkclements — CHANGES_REQUESTED** [review]
exceptional work, thank you! 

couple of points, also would you mind posting benchmarks of main branch vs your branch, and also adding benchmarks that cover this new code (if it's not covered already that is)

**Comment by mcollina** [comment]
Can you also rebase on top of main so we get some CI running? Thanks!

File: lib/modifiers.js (hunk 5 above)
Comment (lrecknagel) [reply lib/modifiers.js:93]: Might be - is fixed

File: lib/modifiers.js (hunk 2 above)
Comment (lrecknagel) [reply lib/modifiers.js:56]: Could you pls explain your point here - I dont get it right now
`subTarget` can be an object, and it works imho 

For the example, which is also in the tests (https://github.com/lrecknagel/fast-redact/blob/multi-wildcard-support-finalization/test/index.js#L1228)

```js
const redactOut = fastRedact({
  paths: ['a[*].x.d[*].p.*.i'],
})
const obj = {
  a: [
    { x: { d: [ { i: undefined, j: 'NR' } ] }},
  ]
}
redactOut(obj)
```

we get the following state for subtarget:

Bildschirmfoto 2022-01-19 um 13 48 31

File: lib/modifiers.js (hunk 1 above)
Comment (lrecknagel) [reply lib/modifiers.js:52]: I think its possible, applies for the example below:

```js
const redact = fastRedact({
  paths: ['a[*].x.d[*].i.*.i']
})
const o = {
  a: [
    { x: { d: [ { i: undefined, j: 'NR' } ] } }
  ]
}
```

Only applies when the given path expects a deeper nested object but is undefined somewhere.

**Review by mcollina — APPROVED** [review]
lgtm on passing CI

**Comment by lrecknagel** [comment]
* Two simple multi-wc benchmarks added.
* See bench of main and the current one:

[001_fast-redact-bench-main.txt](https://github.com/davidmarkclements/fast-redact/files/7897881/001_fast-redact-bench-main.txt)
[002_fast-redact-bench-wc_finalization.txt](https://github.com/davidmarkclements/fast-redact/files/7897882/002_fast-redact-bench-wc_finalization.txt)

Depending on `getObj` which gives a relatively flat object it might be interesting to add another multi-wc bench on deeper nested objects. Would be great if you could express your opinion on that shortly.

**Comment by mcollina** [comment]
I'm good with the benchmarks I'm seeing

**Comment by davidmarkclements** [comment]
thanks @lrecknagel - it all seems good to me, I think let's get this merged, any extra benchmarks that could be PR'd after are always welcome but not demanded

**Pull request merged**